packages/dds/matrix/src/matrix.ts

```diff
@@ -57,6 +59,34 @@ interface ISetOpMetadata {
 	localSeq: number;
 	rowsRefSeq: number;
 	colsRefSeq: number;
+	referenceSeqNumber: number;
+}
+
+export interface ISharedMatrixEvents<T> extends ISharedObjectEvents {
```

Comment:
did you consider alternatives to events, as in general we tell our consumers to not change things from events, and it seems the result of a loss will often be that the loser wants to make some changes somewhere. 

on the same token, i also worry about change the default behavior, i wonder if instead we want a new API, like race cell, which can return an object and allow acting on the win/loss based on the outcome, and then there is no need for this event.

Reply:
We will not change existing DDS in breaking way. That's more of an experimentation (for now), We will need either to fork to have different DDS, or have some policy that essentially allows us to have 2 DDSs in a form of one. I do not think adding new API for existing DDS is a good strategy, as it adds complexity - now we need to define how system works in presence of LWW and FWW operations on same content.

Reply:
The comment about event is valid. That said, I think we essentially gave up on idea to enforce no-reentrancy due to change events. We learned how to live with it (Andrei's work to do partial rebased on pending changes to ensure we have same ref seq number in a batch). We have same problem in consensus DDSs. We could raise events on clean stack. Here it feels Ok, as it impacts only one client. It's worse in CRC, where all clients learn about changes, and doing async means they will learn at different times relative to incoming op stream, which could expose some issues for client.

I really want to get rid of our event system altogether, and rely on coarse-grained events, like - whole document changed, here is new revision. This should help a bit with these kinds of discussions.